File: src/omnia/data/structure/mutable/ArrayList.java
package omnia.data.structure.mutable;

import java.util.Arrays;
import omnia.data.structure.List;

public final class ArrayList<E> extends MaskingList<E> {

  private ArrayList() {
    this(new java.util.ArrayList<>());
  }

  private ArrayList(java.util.ArrayList<E> javaList) {
    super(javaList);
  }

  @SafeVarargs
  public static <E> ArrayList<E> of(E...items) {
    return new ArrayList<>(new java.util.ArrayList<>(Arrays.asList(items)));
  }

  public static <E> ArrayList<E> copyOf(List<? extends E> otherList) {
    ArrayList<E> newList = create();
    otherList.iterator().forEachRemaining(newList::add);
    return newList;
  }

  public static <E> ArrayList<E> create() {
    return new ArrayList<E>();
  }
}

File: src/omnia/data/structure/mutable/ArrayList.kt
package omnia.data.structure.mutable

import omnia.data.structure.List

class ArrayList<E> private constructor(kotlinList: kotlin.collections.List<E> = kotlin.collections.ArrayList()) :
  MaskingList<E>(toArrayList(kotlinList)) {

  companion object {

    @SafeVarargs
    fun <E> of(vararg items: E): ArrayList<E> {
      return ArrayList(mutableListOf(*items))
    }

    @JvmStatic
    fun <E> copyOf(otherList: List<out E>): ArrayList<E> {
      val newList: ArrayList<E> = create()
      otherList.iterator().forEachRemaining { element: E -> newList.add(element) }
      return newList
    }

    @JvmStatic
    fun <E> create(): ArrayList<E> {
      return ArrayList()
    }

    @JvmStatic
    fun <E> toArrayList(list: kotlin.collections.List<E>): kotlin.collections.ArrayList<E> {
      return if (list is kotlin.collections.ArrayList) list else kotlin.collections.ArrayList(list)
    }
  }
}